sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/VersioningTest.groovy

```diff
@@ -0,0 +1,364 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.blob
+
+import com.azure.core.util.Context
+import com.azure.storage.blob.models.BlobListDetails
+import com.azure.storage.blob.models.ListBlobsOptions
+import com.azure.storage.blob.models.PageRange
+import com.azure.storage.blob.models.PublicAccessType
+import com.azure.storage.blob.specialized.PageBlobClient
+import org.apache.commons.lang3.StringUtils
+
+import java.nio.charset.StandardCharsets
+import java.time.Duration
+
+class VersioningTest extends APISpec {
+
+    BlobContainerClient blobContainerClient
+    BlobClient blobClient
+    String blobName
+    String containerName
+    String contentV1 = "contentV1"
+    String contentV2 = "contentV2"
+
+    def setup() {
+        blobName = generateBlobName()
+        containerName = generateContainerName()
+        blobContainerClient = versionedBlobServiceClient.createBlobContainer(containerName)
+        blobClient = blobContainerClient.getBlobClient(blobName)
+    }
+
+    def cleanup() {
+        blobContainerClient.delete();
+    }
+
+    def "Create Block Blob with Version"() {
+        when:
+        def blobItemV1 = blobClient.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize)
+        def blobItemV2 = blobClient.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize, true)
+
+        then:
+        blobItemV1.getVersionId() != null
+        blobItemV2.getVersionId() != null
+        !StringUtils.equals(blobItemV1.getVersionId(), blobItemV2.getVersionId())
+    }
+
+    def "Create Page Blob with Version"() {
+        when:
+        def blobItemV1 = blobClient.getPageBlobClient().create(512)
+        def blobItemV2 = blobClient.getPageBlobClient().create(512, true)
+
+        then:
+        blobItemV1.getVersionId() != null
+        blobItemV2.getVersionId() != null
+        !StringUtils.equals(blobItemV1.getVersionId(), blobItemV2.getVersionId())
+    }
+
+    def "Create Append Blob with Version"() {
+        when:
+        def blobItemV1 = blobClient.getAppendBlobClient().create();
+        def blobItemV2 = blobClient.getAppendBlobClient().create(true);
+
+        then:
+        blobItemV1.getVersionId() != null
+        blobItemV2.getVersionId() != null
+        !StringUtils.equals(blobItemV1.getVersionId(), blobItemV2.getVersionId())
+    }
+
+    def "Retrieve Block Blob by Version"() {
```

Comment:
I think retrieve, properties, and delete are independent of blob type and should share the same code path regardless of blob type, so it may be better to only test that once.

Reply:
Fair point. I'll remove that.